dd-sdk-android/src/main/kotlin/com/datadog/android/core/internal/persistence/file/EncryptedFileHandler.kt

```diff
@@ -0,0 +1,260 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.core.internal.persistence.file
+
+import android.util.Base64
+import com.datadog.android.core.internal.utils.copyTo
+import com.datadog.android.core.internal.utils.devLogger
+import com.datadog.android.log.Logger
+import com.datadog.android.security.Encryption
+import java.io.File
+
+internal class EncryptedFileHandler(
+    internal val encryption: Encryption,
+    internal val delegate: FileHandler,
+    private val internalLogger: Logger,
+    private val base64Encoder: (ByteArray) -> ByteArray = { Base64.encode(it, ENCODING_FLAGS) },
+    private val base64Decoder: (ByteArray) -> ByteArray = {
+        // lambda call is safe-guarded at the call site
+        @Suppress("UnsafeThirdPartyFunctionCall")
+        Base64.decode(it, ENCODING_FLAGS)
+    }
+) : FileHandler {
+
+    // region FileHandler
+
+    override fun writeData(
+        file: File,
+        data: ByteArray,
+        append: Boolean,
+        separator: ByteArray?
+    ): Boolean {
+
+        if (separator != null && !checkSeparator(separator)) {
+            internalLogger.e(INVALID_SEPARATOR_MESSAGE)
+            return false
+        }
+
+        if (append && separator == null) {
+            internalLogger.e(MISSING_SEPARATOR_MESSAGE)
+            return false
+        }
+
+        val encryptedData = encryption.encrypt(data)
+
+        if (data.isNotEmpty() && encryptedData.isEmpty()) {
+            devLogger.e(BAD_ENCRYPTION_RESULT_MESSAGE)
+            return false
+        }
+
+        return delegate.writeData(
+            file,
+            // Base64 produces bytes per US-ASCII encoding, while separator may be in UTF-8 encoding
+            // but this is fine, because UTF-8 is backward compatible with US-ASCII (char in
+            // US-ASCII encoding has the same byte value as char in UTF-8 encoding)
+            base64Encoder(encryptedData),
+            append,
+            separator
+        )
+    }
+
+    override fun readData(
+        file: File,
+        prefix: ByteArray?,
+        suffix: ByteArray?,
+        separator: ByteArray?
+    ): ByteArray {
+
+        if (separator != null && !checkSeparator(separator)) {
+            internalLogger.e(INVALID_SEPARATOR_MESSAGE)
+            return EMPTY_BYTE_ARRAY
+        }
+
+        val data = delegate.readData(file, prefix, suffix, separator)
+
+        if (prefix == null && suffix == null && separator == null) {
+            return decryptSingleItemData(data)
+        }
+
+        val rawData = removeSuffixAndPrefix(data, prefix, suffix)
```

Comment:
this deconstruction could be avoided, but to achieve that encryption should be embedded into the original `FileHandler` and I wanted to keep it responsible only for file manipulation.

Not an expensive operation anyway - array copying is implemented as intrinsic.